### ISSUE TITLE
Predict teleporters and extended teles

### ISSUE THREAD

**Issue by def-** [def-/ddnet#237]
Let's at least predict teles with just 1 destination.

Maybe we can make it deterministic to which destination you get teleported, so that the client can predict it with multiple destinations as well.

**Comment by east** [comment]
That would be nice

> Maybe we can make it deterministic to which destination you get teleported, so that the client can predict it with multiple destinations as well.

A simple solution would be to choose the destination teleporter by gametick but it would be bad if maps rely on non-predictable teleporters.

**Comment by def-** [comment]
The teleporters surely don't feel random right now. By tick would be a good idea, but is that synchronized well enough between client and server?

**Comment by east** [comment]
I'm pretty sure it is. Otherwise the own player prediction wouldn't be that good. I'm not experienced in multiplayer game synchronisation though.

**Comment by cinaera** [comment]
Concerning teleporters: Would be good to have some deterministic solution. There is a race map called `run_button` where you get teleported to 1 of 6 teleporters which heavily influences if you can speedrun the map or not. The original race mod uses the last teleporter always. Some consistency would be nice here. Also helps with correctly predicting the client position. Im not sure if there is any map relying on random teleporting. Couldn't imagine a use-case tbh

**Comment by def-** [comment]
There are definitely maps where it matters that you can't easily predict which tele is going to be taken and so it gets more difficult to catch your partner.

**Comment by CytraL** [comment]
Umm perhaps using a random generator per client (instead of one for all) and use the same seed (for example, CRC of the map + id of the client) ??

**Comment by def-** [comment]
@CytraL But what if the client gets off-sync? Then there's no way to get back to the correct state easily.

**Comment by AssassinTee** [comment]
chances are, that destinations are close to each other. At least in most maps I saw they are close to each other, you can also take this into account and maybe use the average value. A bad prediction is still better than no prediction at all

**Comment by Jupeyy** [comment]
How about we (all maintainers) vote/decide on what exactly we need in a span of ~2 weeks?

1. Do we want teleports to be predicted at all?
2. If 1. was `Yes`, then what kind of change towards teleporters do you want:
 - Full prediction, teleporters should be pseudo random based on a random seed (that is sent to client)
 - Full prediction, teleporters are pseudo random, the server tick is the seed
 - Full prediction, teleporters are not random, humans can easily calcuate which teleporter exit will come (e.g. smth similar to: based on the time in seconds modulo tele exits count)
 - Partial prediction, we keep multiple exits random, but predict single exists. (Inconsistent behavior)
 - I have an even better idea.

I think we should then also accept changes for what we voted (if we want changes).

@Learath2 @def- @Robyt3 @heinrich5991 @Chairn @edg-l @murpii

**Comment by Jupeyy** [comment]
I personally would go with 
1. Yes
2. I am open for:
 - not random (most human friendly, but also bot friendly)
 - pseudo random based on random seed (client can predict, but TAS wouldn't work)

**Comment by Chairn** [comment]
1. Yes
2. Full prediction:
- Pseudo-random and should be reproducible (for speed runners), so based on race time (and not server time)

**Comment by Jupeyy** [comment]
Full prediction with random seed it is then.

The implementation should also inform the teehistorian recorder and increase the teehistorian minor version

**Comment by heinrich5991** [comment]
> isn't much easier than it for botters

Can you explain this? This seems trivially false to me. Given that we had talks about introducing inconsistent physics as an anti-replay bot measure, it seems weird to move into the opposite direction here.

Why I think it's trivially false: The seed tells the bot client which teleporters will be taken on any given run. It can potentially even manipulate that if we increase the seed on every teleport or something.

> Full prediction with random seed it is then.

This still needs a specification of how the randomization should be done, because properties of the bottability and predictability both depend on that.

Should it be stateless and only depend on tick/teleporter position? Should it be stateful per tee? Globally (probably doesn't work because it won't fix the prediction problem)? Etc.

**Comment by Jupeyy** [comment]
> This still needs a specification of how the randomization should be done

Sounds good, are you interested in solving this issue?
I dunno if anyone expressed interest in this issue yet.

**Comment by heinrich5991** [comment]
> Sounds good, are you interested in solving this issue?

No.

I just wanted to say that the decision is incomplete and no one commented with a solution so far:

> Full prediction with random seed it is then.

**Comment by Jupeyy** [comment]
I think that was the biggest hurdle for proper tele prediction. Since opinions on that differ the most.
How the implementation looks like in detail is the job of whoever wants to work on a concept & implementation.
So if you have something in mind that brings this topic further would be appreciated.

**Comment by SollyBunny** [comment]
~~The seed shouldn't be something a bot can control otherwise a single replay can be recorded and the seed can be rerolled.
The seed should also be big enough that a bot can't just brute force rerolling or easily and quickly be rerolled (eg by spam crossing the start line).
The psuedo random number geneator should be "good": evenly distrubuted for obvious reasons and non repeating so that the order can't be remembered (although the cycle would have to be pretty short for that to be possible).~~
This is all assuming that we care if bots can break the psuedo random, once we add it we can't easily change it and picking an expensive and overly complex implementation just causes maintainability issues down the line.
These bots don't yet exist and would require lots of effort to make, probably more than just recording different routes for different teleports and splicing together in real time
If we don't care then having the seed be the tick from some arbritrary data that the server and client can agree on w/o communicating and only generate a new number from this random every time it's needed
An example of what the seed could be composed of would be the join tick and name.

**Comment by heinrich5991** [comment]
> and would require lots of effort to make

Where do you get this from? "I can't think of a way to do it right now" is not really a good argument on that front, I think.

**Comment by SollyBunny** [comment]
> Where do you get this from?

I was talking about extracting the seed which gets harder the more complex the prng is. If we use a simple implementation that is already known, you just have to avoid being able to control the seed or reroll it fast/easily

**Comment by AssassinTee** [comment]
a fix for this should also adress #9269 (probably)

**Comment by sjrc6** [comment]
This is what I would consider to be the most practical teleporter prediction implementation.

There's two main issues to address
1. How to seed projectiles (the client has incomplete knowledge of all projectiles in the game)
2. How much we want to use teleporter randomness as a form of cheat protection, and how hard we should try to do that.

A good seeding for tees and hooks might look like
``[ClientId, GameTick, EntityType, TelePosX, TelePosY, TeeRNGSeed]``

``TeeRNGSeed`` here is an optional new snap object that the server can change to modify the seeding.

The issue with using this seeding for projectiles is that if 2 projectiles enter a teleporter on the same tick they will travel to the same exit, this is rather janky, and probably allows for new gameplay.

The client doesn't know what ID the projectile will receive so we can't simply add the projectile ID to the seed.
We could try to predict what it will be, or wait for the server to tell us before we predict the teleport, but there's a better solution.

If we assume that clients can only fire 1 projectile per tick (which is true since #10528), then the combination of ``[OwnerId, StartTick]`` uniquely identifies all projectiles fired by tees. 

So we can seed projectiles from tees with
``[OwnerId, GameTick, StartTick, EntityType, TelePosX, TelePosY, TeeRNGSeed (of OwnerId)]``

There is a slight problem, which is that entities created by the server could in theory get the same seed if they have the same ``[GameTick, StartTick, EntityType, TelePosX, TelePosY]`` and we won't have a valid TeeRNGSeed. In my opinion this is acceptable, and worth the tradeoff for simplicity. 

Changing the ``TeeRNGSeed`` can potentially cause mispredictions if the client teleports while the updated snap is traveling to them. In my opinion this is acceptable. 

To reduce cheating I suggest we rotate the ``TeeRNGSeed`` on respawn/kill/unspec which won't cause any mispredictions, an aggressive server could rotate this every ~60 seconds, or maybe when the player touches the startline to make it even harder to predict. I think doing it on respawn is good enough for us since it prevents sharing input replay files, this is also not a great place to add this kind of cheat protection because not all maps have a random teleporter.

**Comment by AssassinTee** [comment]
Very good proposal, don't forget a future game tick for the new seed in order to prevent prediction issues on seed switch

**Comment by KebsCS** [comment]
Cheats could move to a desired position and resume replaying the run if a tele out isn't favorable. Overcomplicating the seed just for it to be a barely meaningful change for bottled runs doesn't make much sense. Bots already have huge advantage, if they teleport 1 block further than what's the best, it doesn't make much difference

**Comment by AssassinTee** [comment]
I guess you're right, a minimal PR would be nice for now

**Comment by sjrc6** [comment]
> Cheats could move to a desired position and resume replaying the run if a tele out isn't favorable. Overcomplicating the seed just for it to be a barely meaningful change for bottled runs doesn't make much sense. Bots already have huge advantage, if they teleport 1 block further than what's the best, it doesn't make much difference

I agree but I think adding the server rng seed increases the chances that it gets merged, if we don't do this it will be a slight regression in cheat protection during a time when we have very little. 

Not having any global seed variation would allow regular players to try to time when they enter the teleporter to the global gametick (visible in debug) which would be something mappers need to think about.

I also think preventing naive replay sharing on teleporter maps is a pretty substantial increase in botting protection for not much effort.
